File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Control/UiMenu.ts
import { IControl } from "../UiControl";
import { AveLib, ExtendControlInstance } from "../../AveLib";
import { Window } from "./UiWindow";
import { IconCache, DpiSize, StringKey, DpiSize_2 } from "../UiCommon";

export enum MenuFlag {
	None /**/ = 0x00,
	State /**/ = 0x01,
	Type /**/ = 0x02,
	Text /**/ = 0x04,
	TextSub /**/ = 0x08,
	Key /**/ = 0x10,
	Icon /**/ = 0x20,
	Context /**/ = 0x40,
	Size /**/ = 0x80,
}

export enum MenuType {
	Text,
	Section,
	Separator,
	Content,
}

export enum MenuState {
	Enabled /**/ = 0x00,
	Disabled /**/ = 0x01,
	Checked /**/ = 0x02,
	Radio /**/ = 0x04,
	Default /**/ = 0x08,
	Break /**/ = 0x10,
	NoClose /**/ = 0x20, // Left Click or Keyboard
	NoClose2 /**/ = 0x40, // Right Click
	NoClose3 /**/ = 0x80, // Middle Click
}

export class MenuItem {
	Flag: MenuFlag = MenuFlag.None;
	State: MenuState = MenuState.Enabled;
	Type: MenuType = MenuType.Text;
	Index: number = -1;
	Id: number = 0;
	Icon: IconCache = 0;
	Text: string = "";
	TextSub: string = "";
	Key: string = "";
	Size: DpiSize_2 = new DpiSize_2(DpiSize.FromPixel(0), DpiSize.FromPixel(0));

	constructor(id: number = 0, type: MenuType = MenuType.Text, icon: IconCache = 0, text: string = "", textSub: string = "", key: string = "") {
		this.Id = id;
		this.Type = type;
		this.Icon = icon;
		this.Text = text;
		this.TextSub = textSub;
		this.Key = key;
	}
}

export interface IMenu extends IControl {
	new (window: Window, key?: StringKey): IMenu;

	InsertItem(pItem: MenuItem): number;
	InsertSubMenu(pItem: MenuItem): Menu;
	RemoveById(nId: number): boolean;
	RemoveByIndex(nIndex: number): boolean;
	SetById(pItem: MenuItem): boolean;
	SetByIndex(pItem: MenuItem): boolean;
	GetById(nId: number): MenuItem;
	GetByIndex(nIndex: number): MenuItem;
	GetSubMenuById(nId: number): Menu;
	GetSubMenuByIndex(nIndex: number): Menu;
	Clear(): Menu;
	GetCount(): number;
	SetItemHeight(nHeight: DpiSize): Menu;
	GetItemHeight(): DpiSize;
	SetCheckById(nId: number, b: boolean): number; // return Index
	GetCheckById(nId: number): boolean;
	SetRadioId(nId: number): number; // return Index
	GetRadioId(): number;
	SetGroupRadioId(nIdStart: number, nIdEnd: number, nId: number): number; // return Index, range in [nIdStart, nIdEnd)
	GetGroupRadioId(nIdStart: number, nIdEnd: number): number; // range in [nIdStart, nIdEnd)
	SetContentById(nId: number, pControl: IControl): IControl;
	SetContentByIndex(nIndex: number, pControl: IControl): IControl;
	GetContentById(nId: number): IControl;
	GetContentByIndex(nIndex: number): IControl;

	OnClick(fn: (sender: Menu, nId: number) => void): Menu;
	OnRightClick(fn: (sender: Menu, nId: number) => void): Menu;
	OnMiddleClick(fn: (sender: Menu, nId: number) => void): Menu;
	OnShow(fn: (sender: Menu) => void): Menu;
	OnHide(fn: (sender: Menu) => void): Menu;
}

export class Menu extends (AveLib.UiMenu as IMenu) {
	private m_Content: Set<IControl> = new Set();

	RemoveById(nId: number) {
		this.m_Content.delete(this.GetContentById(nId));
		return super.RemoveById(nId);
	}

	RemoveByIndex(nIndex: number) {
		this.m_Content.delete(this.GetContentByIndex(nIndex));
		return super.RemoveByIndex(nIndex);
	}

	Clear() {
		this.m_Content.clear();
		return super.Clear();
	}

	SetContentById(nId: number, pControl: IControl) {
		const t = super.SetContentById(nId, pControl);
		if (t) this.m_Content.delete(t);
		if (pControl) this.m_Content.add(pControl);
		return t;
	}

	SetContentByIndex(nIndex: number, pControl: IControl) {
		const t = super.SetContentByIndex(nIndex, pControl);
		if (t) this.m_Content.delete(t);
		if (pControl) this.m_Content.add(pControl);
		return t;
	}

	InsertSubMenu(pItem: MenuItem): Menu {
		const menu = super.InsertSubMenu(pItem);
		ExtendControlInstance(menu);
		return menu;
	}
}

File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiMenu.cpp
#include "StdAfx.h"
#include "UiMenu.h"

#define ThisMethod($x) &UiMenu::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiMenu::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiMenu> c_obj;
	}

	void UiMenu::DefineControl()
	{
		AutoAddMethod( InsertItem );
		AutoAddMethod( InsertSubMenu, WrapObjectMix );
		AutoAddMethod( RemoveById );
		AutoAddMethod( RemoveByIndex );
		AutoAddMethod( SetById );
		AutoAddMethod( SetByIndex );
		AutoAddMethod( GetById, WrapObjectGeneric );
		AutoAddMethod( GetByIndex, WrapObjectGeneric );
		AutoAddMethod( GetSubMenuById, WrapObjectMix );
		AutoAddMethod( GetSubMenuByIndex, WrapObjectMix );
		AutoAddMethod( Clear );
		AutoAddMethod( GetCount, WrapObjectGeneric );
		AutoAddMethod( SetItemHeight );
		AutoAddMethod( GetItemHeight, WrapObjectGeneric );
		AutoAddMethod( SetCheckById );
		AutoAddMethod( GetCheckById, WrapObjectGeneric );
		AutoAddMethod( SetRadioId );
		AutoAddMethod( GetRadioId, WrapObjectGeneric );
		AutoAddMethod( SetGroupRadioId );
		AutoAddMethod( GetGroupRadioId, WrapObjectGeneric );
		AutoAddMethod( SetContentById );
		AutoAddMethod( SetContentByIndex );
		AutoAddMethod( GetContentById, WrapObjectGeneric );
		AutoAddMethod( GetContentByIndex, WrapObjectGeneric );

		AutoAddMethod( OnClick       /**/, WrapObjectGeneric );
		AutoAddMethod( OnRightClick  /**/, WrapObjectGeneric );
		AutoAddMethod( OnMiddleClick /**/, WrapObjectGeneric );
		AutoAddMethod( OnShow        /**/, WrapObjectGeneric );
		AutoAddMethod( OnHide        /**/, WrapObjectGeneric );
	}

	U1 UiMenu::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		m_Window = p;

		GetControlTyped().SetIconManager( p->PublicCloneIconManager() );

		return true;
	}

	void UiMenu::__OnClick( Ui::IMenu & sender, U32 nId )
	{
		m_OnClick( this, nId );
	}

	void UiMenu::__OnRightClick( Ui::IMenu & sender, U32 nId )
	{
		m_OnRightClick( this, nId );
	}

	void UiMenu::__OnMiddleClick( Ui::IMenu & sender, U32 nId )
	{
		m_OnMiddleClick( this, nId );
	}

	void UiMenu::__OnShow( Ui::IMenu & sender )
	{
		m_OnShow( this );
	}

	void UiMenu::__OnHide( Ui::IMenu & sender )
	{
		m_OnHide( this );
	}

	S32 UiMenu::InsertItem( const WrapData<UiMenuItem>& pItem )
	{
		Ui::MenuItem mi{};
		WrapDataTransit<Ui::MenuItem>::Transit( mi, pItem );
		return GetControlTyped().InsertItem( mi );
	}

	UiMenu * UiMenu::InsertSubMenu( const MixCallContext & cc, const WrapData<UiMenuItem>& pItem )
	{
		if ( MixCallStage::InUi == cc.GetStage() )
		{
			Ui::MenuItem mi{};
			WrapDataTransit<Ui::MenuItem>::Transit( mi, pItem );
			if ( auto pMenu = GetControlTyped().InsertSubMenu( mi ) )
			{
				pMenu->SetIconManager( m_Window->PublicCloneIconManager() );
				cc.SetUserContext( pMenu );
			}
		}
		else if ( MixCallStage::PostUi == cc.GetStage() )
		{
			if ( !cc.GetUserContext() )
				return nullptr;
			if ( auto r = cc.NewJsObject<UiMenu>() )
			{
				r->SetSharedControl( (Ui::IControl*) cc.GetUserContext() );
				return r;
			}
		}
		return nullptr;
	}

	U1 UiMenu::SetById( const WrapData<UiMenuItem>& pItem )
	{
		Ui::MenuItem mi{};
		WrapDataTransit<Ui::MenuItem>::Transit( mi, pItem );
		return GetControlTyped().SetById( mi );
	}

	U1 UiMenu::SetByIndex( const WrapData<UiMenuItem>& pItem )
	{
		Ui::MenuItem mi{};
		WrapDataTransit<Ui::MenuItem>::Transit( mi, pItem );
		return GetControlTyped().SetByIndex( mi );
	}

	WrapData<UiMenuItem> UiMenu::GetById( U32 nId ) const
	{
		Ui::MenuItem mi{ Ui::MenuItem::Flag_State | Ui::MenuItem::Flag_Type | Ui::MenuItem::Flag_Text | Ui::MenuItem::Flag_TextSub | Ui::MenuItem::Flag_Key | Ui::MenuItem::Flag_Icon | Ui::MenuItem::Flag_Size };
		WChar szText[1024];
		WChar szSubText[1024];
		mi.m_Id = nId;
		mi.m_Text = szText;
		mi.m_TextSub = szSubText;
		mi.m_TextCharLength = CountOf( szText );
		mi.m_TextSubCharLength = CountOf( szSubText );
		if ( !GetControlTyped().GetById( mi ) )
			return {};
		WrapData<UiMenuItem> r;
		r.FromMenuItem( mi );
		return r;
	}

	WrapData<UiMenuItem> UiMenu::GetByIndex( U32 nIndex ) const
	{
		Ui::MenuItem mi{ Ui::MenuItem::Flag_State | Ui::MenuItem::Flag_Type | Ui::MenuItem::Flag_Text | Ui::MenuItem::Flag_TextSub | Ui::MenuItem::Flag_Key | Ui::MenuItem::Flag_Icon | Ui::MenuItem::Flag_Size };
		WChar szText[1024];
		WChar szSubText[1024];
		mi.m_Index = nIndex;
		mi.m_Text = szText;
		mi.m_TextSub = szSubText;
		mi.m_TextCharLength = CountOf( szText );
		mi.m_TextSubCharLength = CountOf( szSubText );
		if ( !GetControlTyped().GetById( mi ) )
			return {};
		WrapData<UiMenuItem> r;
		r.FromMenuItem( mi );
		return r;
	}

	UiMenu * UiMenu::GetSubMenuById( const MixCallContext & cc, U32 nId ) const
	{
		if ( MixCallStage::InUi == cc.GetStage() )
		{
			cc.SetUserContext( GetControlTyped().GetSubMenuById( nId ) );
		}
		else if ( MixCallStage::PostUi == cc.GetStage() )
		{
			if ( !cc.GetUserContext() )
				return nullptr;
			if ( auto r = cc.NewJsObject<UiMenu>() )
			{
				r->SetSharedControl( (Ui::IControl*) cc.GetUserContext() );
				return r;
			}
		}
		return nullptr;
	}

	UiMenu * UiMenu::GetSubMenuByIndex( const MixCallContext & cc, S32 nIndex ) const
	{
		if ( MixCallStage::InUi == cc.GetStage() )
		{
			cc.SetUserContext( GetControlTyped().GetSubMenuByIndex( nIndex ) );
		}
		else if ( MixCallStage::PostUi == cc.GetStage() )
		{
			if ( !cc.GetUserContext() )
				return nullptr;
			if ( auto r = cc.NewJsObject<UiMenu>() )
			{
				r->SetSharedControl( (Ui::IControl*) cc.GetUserContext() );
				return r;
			}
		}
		return nullptr;
	}

	UiMenu* UiMenu::OnClick       /**/( Callback_t       /**/&& fn ) { m_OnClick       /**/ = SetEventCallback<Ui::IMenu::OnClick       /**/>( std::move( fn ), MakeThisFunc( __OnClick       /**/ ) ); return this; }
	UiMenu* UiMenu::OnRightClick  /**/( Callback_t       /**/&& fn ) { m_OnRightClick  /**/ = SetEventCallback<Ui::IMenu::OnRightClick  /**/>( std::move( fn ), MakeThisFunc( __OnRightClick  /**/ ) ); return this; }
	UiMenu* UiMenu::OnMiddleClick /**/( Callback_t       /**/&& fn ) { m_OnMiddleClick /**/ = SetEventCallback<Ui::IMenu::OnMiddleClick /**/>( std::move( fn ), MakeThisFunc( __OnMiddleClick /**/ ) ); return this; }
	UiMenu* UiMenu::OnShow        /**/( OnVisibleChange_t/**/&& fn ) { m_OnShow        /**/ = SetEventCallback<Ui::IMenu::OnShow        /**/>( std::move( fn ), MakeThisFunc( __OnShow        /**/ ) ); return this; }
	UiMenu* UiMenu::OnHide        /**/( OnVisibleChange_t/**/&& fn ) { m_OnHide        /**/ = SetEventCallback<Ui::IMenu::OnHide        /**/>( std::move( fn ), MakeThisFunc( __OnHide        /**/ ) ); return this; }
}
